### PR TITLE
Protect OOProc replay loop from Activity exceptions

### PR DESCRIPTION
**Pull request by davidmrdavid**
We recently introduced a new OOProc replay procedure that more accurately represents composite tasks (WhenAll/WhenAny) in the extension.

In the **old** approach, **all** OOProc SDK actions would be awaited **inside** a `WhenAny` Task, which ran the risk of creating a race condition where not all tasks were scheduled at the end of the replay loop. In the new replay procedure, we instead await each task one at a time, avoiding this race condition.

However, in doing so, we've lost the ability to try-catch Activity exceptions in orchestrators. This is because awaiting a failed Task will throw an exception in the replay loop whereas, before, that exception would be "swallowed" by the surrounding `WhenAny` task.

This PR simply introduces a `WhenAny` wrapper to awaited tasks in the OOProc Shim, so that the replay procedure can continue in the presence of errors.

### Issue describing the changes in this PR

N/A

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [ ] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk

### CODE REVIEW

**Review by ConnorMcMahon — APPROVED** [review]
Approved with suggestions.

File: src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs:204]: I'm willing to block the PR on this because this is extremely roundabout way to swallow an exception, and very confusing. +1 to doing try/catch instead.

Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs:204]: Fair enough, I only restored the `Task.WhenAny` because that recovers the "old and tested" way in which we swallowed these exceptions. A part of me wanted to bring it back in case there were even **more** **unknown** reasons why `WhenAny` was chosen in the past, but I agree that a try-catch is better for this particular situation.

Happy to change this

Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Listener/OutOfProcOrchestrationShim.cs:204]: `Task.WhenAny` was used to work for the case when we received multiple actions in an array from a .All() or .Any() call from the out-of-proc layer.

However, that clearly was incorrect for the .All() case, which lead to nondeterminism problems occasionally, as well as  issues with nested compound tasks.

There is no reason to use Task.WhenAny() now that we properly translate the tasks.

**Review by cgillum — APPROVED** [review]
The latest iteration addresses my main concern. Thanks!

**Comment by davidmrdavid** [comment]
@ConnorMcMahon, there should be no reason to backport this to V1, right? Just checking before merging

**Pull request merged**
